File: workspace/random_pairing.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Oct  7 13:58:26 2025

@author: matthew
"""

import sys
import os
import random


# %% Read input file

if len(sys.argv) <= 1:
    print(
        'Invalid arguments, please include a file name for the generator.\n\n'
        'The file can be a vertical list of first names, or a vertical list '
        'of names in the form\n'
        '<Last Name>, <First Name>\n\n'
        'For example:\n'
        'carl\n'
        'katia\n'
        'lucia\n\n'
        'or\n\n'
        'warburton, carl\n'
        'grim, katia\n'
        'mar, lucia\n'
        '\n'
        'Usage: $ random_pairing.py <filename>'
    )
    sys.exit()
if os.path.exists(sys.argv[1]):
    with open(sys.argv[1]) as file:
        raw_lines = file.readlines()
        name_lines = [line for line in raw_lines if line.strip()]
else:
    print('File not found, exiting...')
    sys.exit()


# %% Parse input file
names = []

"""
print('Are the names in <Last name>, <First name> format? (y/N)')
try:
    format_q = input('> ').lower()
except KeyboardInterrupt:
    print('Exiting...')
    sys.exit()
"""

if ',' in name_lines[0]:
    for line in name_lines:
        first_name = line.split(',')[1].strip()
        last_name = line.split(',')[0].strip()
        names.append(f'{first_name} {last_name}')
else:
    for line in name_lines:
        names.append(line.strip())

random.shuffle(names)


# %% Generate and print the pairs

pairs = []

# Set up a trio if there are an odd number of students
if len(names) % 2 == 1:
    tmp = (
        names.pop().title(),
        names.pop().title(),
        names.pop().title()
    )
    pairs.append(tmp)

# Pair up the students and add them to the list of pairs
while names:
    tmp = (names.pop().title(), names.pop().title())
    pairs.append(tmp)

# Make a list that looks nicer to output
output_list = []
tmp_string = ''
for pair in pairs:
    output_list.append(' and '.join(pair))

# Make one big string to print and potentially save to a file
output_string = '\n'.join(output_list)
print('\n\n' + output_string + '\n\n')


# %% Ask to save the file

print('Would you like to save the list to a file? (y/N)')
try:
    save_file = input('> ').lower()
except KeyboardInterrupt:
    print('Exiting...')
    sys.exit()
if save_file in ('y', 'yes'):
    print('Enter a file name for the list.'
          ' Ex. student_pairs')
    try:
        file_name = input('> ') + '.txt'
    except KeyboardInterrupt:
        print('Exiting...')
        sys.exit()

    if os.path.exists(file_name):
        overwrite = input('File already exists, overwrite? (y/N)\n> ').lower()
        if overwrite in ('y', 'yes'):
            print(f'Overwriting list to {file_name}...')
            with open(file_name, 'w') as f:
                f.write(output_string)
        else:
            print('File not written, exiting...')
            sys.exit()
    else:
        print(f'Writing list to {file_name}...')
        with open(file_name, 'w') as f:
            f.write(output_string)
    print('Exiting...')
else:
    print('File not written, exiting...')
    sys.exit()

File: workspace/random_pairing.lua
#!/usr/bin/env lua

--[[ Read the input file ]]
if #arg < 1 then
	error([[
		
		Invalid arguments, please include a file name for the generator
		The file can be a vertical list of first names, or a vertical list 
		of names in the form: <Last Name>, <First Name>

		For example:
		carl
		katia
		lucia

		or

		warburton, carl
		grim, katia
		mar, lucia
		
		Usage: $ random_pairing.py <filename>
	]])
end

local function split(str, delim)
	--[[ Splits a string into a table based on delim ]]
	local t = {}
	delim = delim or "%s+" -- default to whitespace
	for token in string.gmatch(str, "[^" .. delim .. "]+") do
		token = token:gsub('%s', '')
		table.insert(t, token)
	end
	return t
end

local function shuffle(t)
	local j, temp
	--[[ Start at the end and work backwards ]]
	for i = #t, 1, -1 do
		j = math.random(i) -- Get a random number bounded by i
		temp = t[i] -- Store the value at i
		t[i] = t[j] -- Swap the value at j to i
		t[j] = temp -- Place the value stored for i at j
	end
end

local function title(s)
	return string.gsub(s, '%f[%w]%l', string.upper)
end

file = io.open(arg[1])

names = {}

if file then
	for line in file:lines() do
		--[[ Filter out empty lines ]]
		if not line:match('%S') then -- %S is the inverse of whitespace characters
			goto continue
		end
		--[[ If the line is in the form <last name>, <first name> ]]
		if line:match(',') then
			local parts = split(line, ',') -- strip whitespace
			local first_name = parts[2]
			local last_name = parts[1]
			table.insert(names, first_name .. ' ' .. last_name)
		--[[ If the line is in the form of <first name> ]]
		else
			name = line:gsub('%s', '') -- strip whitespace
			table.insert(names, name)
		end

		::continue::
	end
else
	error('Error: file not found!')
end

--[[ Randomize the table ]]
shuffle(names)

--[[ Make the table of pairs. Each pair is a table ]]
pairs = {}

if #names % 2 == 1 then
	local temp = {}
	for i = 1, 3 do
		table.insert(temp, table.remove(names, #names))
	end
	table.insert(pairs, temp)
end

while #names > 0 do
	local temp = {}
	for i = 1, 2 do
		table.insert(temp, table.remove(names, #names))
	end
	table.insert(pairs, temp)
end

--[[ Make a table of strings to print ]]
output_table = {}

for _,pair in ipairs(pairs) do
	for i = 1, #pair do
		pair[i] = title(pair[i])
	end
	local out = table.concat(pair, ' and ')
	table.insert(output_table, out)
end

for _,v in ipairs(output_table) do
	print(v)
end

--[[ Saving to a file ]]
io.write('Would you like to save the pairings to a file? [y/N]\n> ') -- We use this instead of print() to avoid printing a newline char
io.flush()
response = io.read()
response = response:gsub('%u', string.lower)

if response:match('^y') then
	if arg[2] then
		out_file = arg[2]
	else
		io.write('Please enter a file name: ')
		io.flush()
		out_file = io.read()
	end

	--[[ Try to open the file to see if it exists ]]
	output = io.open(out_file)
	if output then
		io.close(output) -- close the file
		print('File already exists, overwrite? [Y/n]')
		io.write('> ')
		io.flush()
		overwrite = io.read()
		overwrite = overwrite:gsub('%u', string.lower)
		if overwrite:match('^n') then
			print('File not written.')
			os.exit()
		else
			--[[ Open and close the file in write mode to clear it ]]
			output = io.open(out_file, 'w')
			io.close(output)
		end
	end

	--[[ Open the file in write mode ]]
	output = io.open(out_file, 'a')
	for _, pairing in ipairs(output_table) do
		output:write(pairing .. '\n')
	end
	io.close(output)
	print('Pairings written to file ' .. out_file .. '.')
end
